src/dns/coredns.rs

```diff
@@ -158,6 +158,57 @@ impl CoreDns {
                                 self.kill_switch.lock().unwrap()
                             );
 
+                            // if record type is PTR try resolving early and return if record found
+                            if record_type == RecordType::PTR {
+                                let mut ptr_lookup_ip: String;
+                                // Are we IPv4 or IPv6?
+                                if name.contains(".in-addr.arpa.") {
+                                    // IPv4
+                                    ptr_lookup_ip = name.trim_end_matches(".in-addr.arpa.").split('.').rev().collect::<Vec<&str>>().join(".");
+                                } else if name.contains(".ip6.arpa.") {
+                                    // IPv6
+                                    ptr_lookup_ip = name.trim_end_matches(".ip6.arpa.").split('.').rev().collect::<String>();
+                                    // We removed all periods; now we need to insert a : every 4 characters.
+                                    // split_off() reduces the original string to 4 characters and returns the remainder.
+                                    // So append the 4-character and continue going until we run out of characters.
+                                    let mut split: Vec<String> = Vec::new();
+                                    while ptr_lookup_ip.len() > 4 {
+                                        let tmp = ptr_lookup_ip.split_off(4);
+                                        split.push(ptr_lookup_ip);
+                                        ptr_lookup_ip = tmp;
+                                    }
+                                    // Length should be equal to 4 here, but just use > 0 for safety.
+                                    if ptr_lookup_ip.len() > 0 {
+                                        split.push(ptr_lookup_ip);
+                                    }
+                                    ptr_lookup_ip = split.join(":");
+                                } else {
+                                    ptr_lookup_ip = String::from("");
+                                }
+
+                                trace!("perform lookup reverse lookup for ip: {:?}", ptr_lookup_ip.to_owned());
+                                // We should probably log malformed queries, but for now if-let should be fine.
+                                if let Ok(lookup_ip) = ptr_lookup_ip.parse() {
+                                    let reverse_lookup = self.backend.reverse_lookup(&src_address.ip(), &lookup_ip);
+                                    if !reverse_lookup.is_empty() {
```

Comment:
This seems like an argument for `fn reverse_lookup() -> Option<Vec<String>>`.


Then inside the function instead of returning an empty vec it's just `None`, and this is an `if let Some()` etc.

Reply:
Fair. I'd also still like to make it return &str. But these are both things for Monday.

Reply:
(For a final return type of `Option<Vec<&str>>`)

Reply:
>  But these are both things for Monday.

Definitely, just glanced at the code out of curiosity; nothing I'm saying here should be considered to block anything!  And have a good weekend!